### PR TITLE
scripts: pylib: twister: testplan: Fix Issue #65477

### LINKED ISSUE #65477: Testplan platform key checking may not work at all
**Describe the bug**
Lines 845-870 of the testplan module describe our approach to filtering test platforms with regard to specified platform key fields.

However, the code seems to be incorrect. The explanatory text is describing one approach, the variable names suggest a different approach and the code itself does something else entirely, checking whether a testsuite platform key list has `None`s in it or `'na'`s. From context, I think it is a bug and instead the code should check whether given platform has all fields specified in the testsuite.

**To Reproduce**
N/A

**Expected behavior**
Platform key checking should verify that all required platform keys exist on a given platform.

**Impact**
Platform key functionality seems to be effectively missing.

**Logs and console output**
N/A

**Environment (please complete the following information):**
 - OS: Linux
 - 26fd55e0a750305fbd5d9db750225d7e707458ae

**Additional context**
https://github.com/zephyrproject-rtos/zephyr/blob/6b87eedee02020392270305fac8258557598b074/scripts/pylib/twister/twisterlib/testplan.py#L845-L870
A PR is currently in the works.

### PR DESCRIPTION
**Pull request by LukaszMrugala**
Fixes #65477
Platform key checking seemed to be erroneous; now the variable names, comments and code seem in line with each other.

~~Modified code now follows the 80-char-line standard.~~

### CODE REVIEW

**Comment by thedjnK** [comment]
> Modified code now follows the 80-char-line standard.

This is not a zephyr rule, the zephyr rule is 100 char lines

**Review by hakehuang — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]
Looks fine to me

**Comment by nashif** [comment]
> Modified code now follows the 80-char-line standard.

please do not mix such changes with the actual functional changes, this makes the PR difficult to review. And as said before 80 char-line is not the standard.

**Comment by LukaszMrugala** [comment]
Reverted the line length changes.

File: scripts/pylib/twister/twisterlib/testplan.py [hunk 1]
```diff
@@ -842,32 +842,35 @@ def apply_filters(self, **kwargs):
...
-                        keyed_test = keyed_tests.get(test_key)
+                    keys = [getattr(plat, key_field) for key_field in key_fields]
+                    for key in keys:
+                        if key is None or key == 'na':
+                            instance.add_filter(
+                                f"Excluded platform missing key fields demanded by test {key_fields}",
+                                Filters.PLATFORM
+                            )
+                            break
+                    else:
```
Comment (golowanow) [inline scripts/pylib/twister/twisterlib/testplan.py:861]: the `else:` block - is one tabulation level missed ?

**Review by golowanow — DISMISSED** [review]

File: scripts/pylib/twister/twisterlib/testplan.py [hunk 2]
```diff
@@ -842,32 +842,35 @@ def apply_filters(self, **kwargs):
...
+                        test_keys = copy.deepcopy(keys)
+                        test_keys.append(ts.name)
+                        test_keys = tuple(test_keys)
+                        keyed_test = keyed_tests.get(test_keys)
                         if keyed_test is not None:
                             plat_key = {key_field: getattr(keyed_test['plat'], key_field) for key_field in key_fields}
-                            instance.add_filter(f"Excluded test already covered for key {tuple(key)} by platform {keyed_test['plat'].name} having key {plat_key}", Filters.PLATFORM_KEY)
+                            instance.add_filter(
+                                f"Excluded test already covered for key {tuple(key)} by platform {keyed_test['plat'].name} having key {plat_key}",
```
Comment (nashif) [inline scripts/pylib/twister/twisterlib/testplan.py:869]: this ends up as:

```"reason":"Excluded test already covered for key ('q', 'e', 'm', 'u') by platform qemu_xtensa_mmu having key {'simulation': 'qemu'}"```

File: scripts/pylib/twister/twisterlib/testplan.py [hunk 3]
```diff
@@ -842,32 +842,35 @@ def apply_filters(self, **kwargs):
...
+                                Filters.PLATFORM_KEY
+                            )
                         else:
-                            keyed_tests[test_key] = {'plat': plat, 'ts': ts}
-                    else:
-                        instance.add_filter(f"Excluded platform missing key fields demanded by test {key_fields}", Filters.PLATFORM)
+                            keyed_tests[test_keys] = {'plat': plat, 'ts': ts}
```
Comment (nashif) [inline scripts/pylib/twister/twisterlib/testplan.py:873]: an issue that existed in original implementation: we need to check if that platform was filtered and should not use it if it was filtered, so 

```
-                            keyed_tests[test_key] = {'plat': plat, 'ts': ts}
+                            if not instance.filters:
+                                keyed_tests[test_key] = {'plat': plat, 'ts': ts}

```

**Review by nashif — CHANGES_REQUESTED** [review]

File: scripts/pylib/twister/twisterlib/testplan.py (hunk 3 above)
Comment (nashif) [reply scripts/pylib/twister/twisterlib/testplan.py:873]: see https://github.com/zephyrproject-rtos/zephyr/pull/65834

File: scripts/pylib/twister/twisterlib/testplan.py (hunk 2 above)
Comment (teburd) [reply scripts/pylib/twister/twisterlib/testplan.py:869]: ew yeah please fix

File: scripts/pylib/twister/twisterlib/testplan.py (hunk 3 above)
Comment (LukaszMrugala) [reply scripts/pylib/twister/twisterlib/testplan.py]: Incorporated #65834's changes.

File: scripts/pylib/twister/twisterlib/testplan.py (hunk 2 above)
Comment (LukaszMrugala) [reply scripts/pylib/twister/twisterlib/testplan.py]: Changed to plain `{key}`.

**Comment by nashif** [comment]
you need to fix the commit messages, do not mention bug numbers in the commit title, document what you are changing instead.
When you incorportate changes from other PR and other people, please cherry pick them instead and leave original commit and author in place.

**Comment by nashif** [comment]
Your commit message still says:

`scripts: pylib: twister: testplan: Fix Issue #65477`

**Review by nashif — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**